polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -75,19 +76,25 @@ static bool isValidMemRefType(Type type) {
   return true;
 }
 
-// Returns true if the block of \p startOp contains an instruction (after \p
+// Returns true if the block of \p startOp contains an operation (after \p
 // startOp) that has a side effects on \p val.
-static auto existsSideEffectAfter(Value val, Operation *startOp) {
+static bool existsSideEffectAfter(Value val, Operation *startOp,
+                                  AliasAnalysis &aliasAnalysis) {
   assert(startOp && "Expecting a valid pointer");
   Block *block = startOp->getBlock();
 
   for (Operation &op : *block) {
     if (op.isBeforeInBlock(startOp) || isMemoryEffectFree(&op))
       continue;
 
-    // Currently bail out if any subsequent operation has side effects.
-    // TODO: check whether the operation alias with 'val'.
-    return true;
+    // Check whether the operation uses an operand that alias with 'val'.
+    if (auto MEI = dyn_cast<MemoryEffectOpInterface>(op)) {
+      if (!llvm::all_of(op.getOperands(), [&](Value operand) {
+            AliasResult aliasRes = aliasAnalysis.alias(operand, val);
+            return aliasRes.isNo();
+          }))
+        return true;
+    }
```

Comment:
Handle the case in which `op` is not a `MemoryEffectOpInterface`, but has the `HasRecursiveMemoryEffects` trait.